### ISSUE TITLE
Balanced sampling for classification problems

### ISSUE THREAD

**Issue by dsherry** [alteryx/evalml#973]
#971 tracks adding a data check to detect and warn about severe class-imbalance. #457 tracks adding an automl objective which can perform well on imbalanced datasets

Detecting and warning about class imbalance could be helpful, but ideally we could improve our classifier quality by down-sampling the training data to flatten the distribution of the target over each class.

There's a couple levels this problem exists at:
1. In automl, if we want to make a fair comparison between models, we should either use an automl objective which is hardened to imbalanced data (like Neyman-Pearson tracked by #457), or we should balance the data before computing the objective score.
2. During the training of each pipeline, each CV fold should be down-sampled to have a balanced distribution of labels. A simple way to do this is to compute the splits in automl as we do now, apply down-sampling to balance, and then train/evaluate each pipeline on each split. We could also have the pipelines generate the splits and down-sample internally, but if this is done right there wouldn't be any difference in accuracy.

I see [sklearn's doc mentions class imbalance](https://scikit-learn.org/stable/modules/cross_validation.html#cross-validation-iterators-with-stratification-based-on-class-labels), but I don't agree with their recommendation "to ensure that relative class frequencies is approximately preserved in each train and validation fold." I suppose we could put this to the test and see which strategy seems to generalize more effectively.

**Comment by jeremyliweishih** [comment]
After discussion with @dsherry on Friday and preliminary analysis using performance tests, we've decided to push this back until we have larger and more representative datasets within looking glass to make a decision on #1266 even though the analysis showed good results.

**Comment by jeremyliweishih** [comment]
Moving back to backlog.

**Comment by rpeck** [comment]
@dsherry I've gone back a few times to your last paragraph in your first comment, about `sklearn` suggesting maintaining class balance in the CV folds.

I think there's a balance. If the class imbalance is way out of whack for some fold(s), because the minority class is so small, then there's not much to learn from regarding the minority class in those folds, and the performance variance on the minority class across the folds will be terrible.

OTOH, there's a regularization aspect to the samples for the folds, similar to the regularization aspect of the column sampling in Random Forest, or the regularization done by bagging. 

I think if we resample (down/up) before doing xval we'll be well covered.
